packages/purpose-platformstate-writer/src/utils.ts

```diff
@@ -97,6 +131,403 @@ export const deletePlatformPurposeEntry = async (
   await dynamoDBClient.send(command);
 };
 
+export const readTokenEntriesByGSIPKPurposeId = async (
+  dynamoDBClient: DynamoDBClient,
+  purposeId: PurposeId,
+  exclusiveStartKey?: Record<string, AttributeValue>
+): Promise<{
+  tokenStateEntries: TokenGenerationStatesClientPurposeEntry[];
+  lastEvaluatedKey?: Record<string, AttributeValue>;
+}> => {
+  const input: QueryInput = {
+    TableName: config.tokenGenerationReadModelTableNameTokenGeneration,
+    IndexName: "Purpose",
+    KeyConditionExpression: `GSIPK_purposeId = :gsiValue`,
+    ExpressionAttributeValues: {
+      ":gsiValue": { S: purposeId },
+    },
+    ExclusiveStartKey: exclusiveStartKey,
+  };
+  const command = new QueryCommand(input);
+  const data: QueryCommandOutput = await dynamoDBClient.send(command);
+
+  if (!data.Items) {
+    throw genericInternalError(
+      `Unable to read token state entries: result ${JSON.stringify(data)} `
+    );
+  } else {
+    const unmarshalledItems = data.Items.map((item) => unmarshall(item));
+
+    const tokenStateEntries = z
+      .array(TokenGenerationStatesClientPurposeEntry)
+      .safeParse(unmarshalledItems);
+
+    if (!tokenStateEntries.success) {
+      throw genericInternalError(
+        `Unable to parse token state entry item: result ${JSON.stringify(
+          tokenStateEntries
+        )} - data ${JSON.stringify(data)} `
+      );
+    }
+
+    return {
+      tokenStateEntries: tokenStateEntries.data,
+      lastEvaluatedKey: data.LastEvaluatedKey,
+    };
+  }
+};
+
+export const readAllTokenEntriesByGSIPKPurposeId = async (
+  dynamoDBClient: DynamoDBClient,
+  purposeId: PurposeId
+): Promise<TokenGenerationStatesClientPurposeEntry[]> => {
+  const runPaginatedQuery = async (
+    dynamoDBClient: DynamoDBClient,
+    purposeId: PurposeId,
+    exclusiveStartKey?: Record<string, AttributeValue>
+  ): Promise<TokenGenerationStatesClientPurposeEntry[]> => {
+    const result = await readTokenEntriesByGSIPKPurposeId(
+      dynamoDBClient,
+      purposeId,
+      exclusiveStartKey
+    );
+    if (!result.lastEvaluatedKey) {
+      return result.tokenStateEntries;
+    } else {
+      return [
+        ...result.tokenStateEntries,
+        ...(await runPaginatedQuery(
+          dynamoDBClient,
+          purposeId,
+          result.lastEvaluatedKey
+        )),
+      ];
+    }
+  };
+
+  return await runPaginatedQuery(dynamoDBClient, purposeId);
+};
+
+export const updatePurposeDataInPlatformStatesEntry = async ({
+  dynamoDBClient,
+  primaryKey,
+  purposeState,
+  version,
+  purposeVersionId,
+}: {
+  dynamoDBClient: DynamoDBClient;
+  primaryKey: PlatformStatesPurposePK;
+  purposeState: ItemState;
+  version: number;
+  purposeVersionId?: PurposeVersionId;
```

Comment:
Updated. That change made the logic of the following events the same because `purposeState` is based on  the presence of an active `purposeVersion` in `purpose.versions`:
```
      { type: "NewPurposeVersionActivated" },
      { type: "PurposeVersionActivated" },
      { type: "PurposeVersionSuspendedByConsumer" },
      { type: "PurposeVersionSuspendedByProducer" },
      { type: "PurposeVersionUnsuspendedByConsumer" },
      { type: "PurposeVersionUnsuspendedByProducer" },
```